### PR TITLE
Stable sort for (data/newtype) instances

### PR DESCRIPTION
**Pull request by alexbiehl**
@hsyl20 Found non determinism in Haddock output. Obviously we were comparing instance heads by Unique. The effect is probably aggravated by the Parallel Haddock work. 

https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4951#note_329145

### CODE REVIEW

**Review by Kleidukos — APPROVED** [review]
LGTM

File: haddock-api/src/Haddock/Interface/AttachInstances.hs [hunk 1]
```diff
@@ -194,7 +195,8 @@ instance Ord SName where
 -- For the benefit of the user (looks nice and predictable) and the
 -- tests (which prefer output to be deterministic).
 data SimpleType = SimpleType SName [SimpleType]
```
Comment (Kleidukos) [inline haddock-api/src/Haddock/Interface/AttachInstances.hs:197]: Any reason why you're not using `!` and `{-# UNPACK #-}` for the constructors?

Comment (alexbiehl) [reply haddock-api/src/Haddock/Interface/AttachInstances.hs:197]: No reason, it's been non-strict before so I thought I'd just leave it.

Comment (Kleidukos) [reply haddock-api/src/Haddock/Interface/AttachInstances.hs:197]: Alright. :)

**Pull request merged**